cranelift/bitset/src/compound.rs

```diff
@@ -0,0 +1,471 @@
+//! Compound bit sets.
+
+use crate::scalar::{self, ScalarBitSet};
+use alloc::{vec, vec::Vec};
+use core::mem;
+
+/// A large bit set backed by dynamically-sized storage.
+///
+/// # Example
+///
+/// ```
+/// use cranelift_bitset::CompoundBitSet;
+///
+/// // Create a new bitset.
+/// let mut bitset = CompoundBitSet::new();
+///
+/// // Bitsets are initially empty.
+/// assert!(bitset.is_empty());
+/// assert_eq!(bitset.len(), 0);
+///
+/// // Insert into the bitset.
+/// bitset.insert(444);
+/// bitset.insert(555);
+/// bitset.insert(666);
+///
+/// // Now the bitset is not empty.
+/// assert_eq!(bitset.len(), 3);
+/// assert!(!bitset.is_empty());
+/// assert!(bitset.contains(444));
+/// assert!(bitset.contains(555));
+/// assert!(bitset.contains(666));
+///
+/// // Remove an element from the bitset.
+/// let was_present = bitset.remove(666);
+/// assert!(was_present);
+/// assert!(!bitset.contains(666));
+/// assert_eq!(bitset.len(), 2);
+///
+/// // Can iterate over the elements in the set.
+/// let elems: Vec<_> = bitset.iter().collect();
+/// assert_eq!(elems, [444, 555]);
+/// ```
+#[derive(Clone, PartialEq, Eq)]
+#[cfg_attr(
+    feature = "enable-serde",
+    derive(serde_derive::Serialize, serde_derive::Deserialize)
+)]
+pub struct CompoundBitSet {
+    elems: Vec<ScalarBitSet<usize>>,
+    len: usize,
+}
+
+impl core::fmt::Debug for CompoundBitSet {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        write!(f, "CompoundBitSet ")?;
+        f.debug_set().entries(self.iter()).finish()
+    }
+}
+
+impl Default for CompoundBitSet {
+    #[inline]
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+const BITS_PER_WORD: usize = mem::size_of::<usize>() * 8;
+
+impl CompoundBitSet {
+    /// Construct a new, empty bit set.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use cranelift_bitset::CompoundBitSet;
+    ///
+    /// let bitset = CompoundBitSet::new();
+    ///
+    /// assert!(bitset.is_empty());
+    /// ```
+    #[inline]
+    pub fn new() -> Self {
+        CompoundBitSet {
+            elems: vec![],
+            len: 0,
+        }
+    }
+
+    /// Construct a new, empty bit set with space reserved to store any element
+    /// `x` such that `x < capacity`.
+    ///
+    /// The actual capacity reserved may be greater than that requested.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use cranelift_bitset::CompoundBitSet;
+    ///
+    /// let bitset = CompoundBitSet::with_capacity(4096);
+    ///
+    /// assert!(bitset.is_empty());
+    /// assert!(bitset.capacity() >= 4096);
+    /// ```
+    #[inline]
+    pub fn with_capacity(capacity: usize) -> Self {
+        // Divide `capacity` by `BITS_PER_WORD`, rounding up rather than down.
+        let elems_cap = capacity + BITS_PER_WORD - 1 / BITS_PER_WORD;
```

Comment:
I think this might need `(...)` to disambiguate?

Reply:
Whoops!